### ISSUE TITLE
clear up confusion about `outgoing`

### ISSUE THREAD

**Issue by e0** [MetabolicAtlas/MetabolicAtlas#1227]
**Please describe the feature**

As a user, I want to clearly understand how the API response for related reactions matches what is shown on the website so that I am not confused by the different forms of data representation.

See **Additional context** below for more details. If the explanation is correct, the Documentation should be updated accordingly. Otherwise please discuss in the comments about how we should best move forward with this issue.

**Acceptance Criteria**

- [ ] The web page documentation is updated accordingly
- [ ] The changes are in alignment with the privacy policy
- [ ] The changes are tested on multiple models and browsers
- [ ] The API response is the same as before
- [ ] The changes do not affect the deployment
- [ ] The changes do not reduce the performance

**Additional context**

There seem to be two main questions, or points of confusion, revolving the `outgoing` field from the API response.
1. What is the relationship of `outgoing` of a metabolite and the directionality of the associated reaction?
2. If `outgoing` is `true` for a metabolite, does that mean it is a reactant (left side of equation) or product (right side of equation) in a reaction?

### Question 1

> _What is the relationship of `outgoing` of a metabolite and the directionality of the associated reaction?_

The example that was originally mentioned for this issue is from the following page: `/explore/Human-GEM/gem-browser/metabolite/MAM02191m`. In the related reactions table, we have the following reaction and equation.

<img width="666" alt="image" src="https://user-images.githubusercontent.com/423498/203503941-541f90ec-ac7f-4b06-bb95-faa146300ea9.png">

In the accompanying API response (`/api/v2/metabolites/MAM02191m/related-reactions?model=HumanGem&version=1_12_0&limit=1000&isForAllCompartments=false`), the response for reaction `MAR08662` looks like the following. 

![image](https://user-images.githubusercontent.com/423498/203504652-ba520c36-6690-44db-b5c2-d23fb2416fd9.png)

#### Answer
The `reversible` field for a reaction (indicated by the arrow) is what decides if the reaction is bidirectional. The `outgoing` field for a metabolite indicates which side of the equation it is on and has nothing to do with if it is bidirectional or not. Since this is a bidirectional reaction, the "which side (if a metabolite is product or reactant) question" is not important because they are both a product and a reactant. What is important is that there the two groups of metabolites on opposite sides, which is the purpose of the `outgoing` field for metabolites in bidirectional reactions.

### Question 2
> _If `outgoing` is `true` for a metabolite, does that mean it is a reactant (left side of equation) or product (right side of equation) in a reaction?_

It seems that it's not straightforward what `outgoing` implies. In fact it may intuitively seem that if `outgoing` is `true` for a metabolite, then it is a reaction product (right side of equation).

I think this is where there could a mismatch between what seems natural in thinking about graph database design and what seems intuitive for researchers. To illustrate this from the database design perspective, in the following image, it shows the metabolites (orange nodes) involved in a reaction (green node). The reaction is `MAR04837` (`H2O [e] + trehalose [e] ⇒ 2 glucose [e]`) and the selected metabolite node is `MAM02040e` (`H2O`). In this example, the `true` value of `outgoing` from `H2O` is derived from the fact that it **_has_** an outgoing edge to the reaction, as opposed to `H2O` **_is_** the resulting outgoing node.

<img width="841" alt="image" src="https://user-images.githubusercontent.com/423498/203523115-9f8cbcfe-e868-4dbd-8e10-57003c1b7d4f.png">

The API response for this reaction looks like the following:

<img width="716" alt="image" src="https://user-images.githubusercontent.com/423498/203525663-6ca1265c-7310-402d-b401-a4141dce2f26.png">

#### Answer
When `outgoing` is `true` for a metabolite, it means that it is a reactant.
When `outgoing` is `false` for a metabolite, it means that it is a product.

#### Discussion
If it seems like the reverse (`outgoing` representing a product) is more intuitive, should we change anything in the API response? If not then this should be clearly documented to avoid confusion.

A suggestion was made that the `stoichiometry` value could be used as well. I do not have enough domain knowledge about if this interfere with values from `data-files` and other possible implications. It would be great to get some other perspective on this.

**Comment by e0** [comment]
Just want to ping @haowang-bioinfo and @JonathanRob who were involved in the initial discussion so it's easier for you to spot this issue.

**Comment by haowang-bioinfo** [comment]
@e0 thank you very much for preparing the impressively detail issue for discussion!

> If it seems like the reverse (outgoing representing a product) is more intuitive, should we change anything in the API response? If not then this should be clearly documented to avoid confusion.

From a personal perspective, it might be to better simply remove "outgoing" column from the API response, due to the confusion it brings.

> A suggestion was made that the stoichiometry value could be used as well. I do not have enough domain knowledge about if this interfere with values from data-files and other possible implications. It would be great to get some other perspective on this.

Yes, this seems to be a very good suggestion, i.e. directly using the stoichiometry coefficient values, in which positive/negative state actually carries the information in "outgoing" column

**Comment by JonathanRob** [comment]
Thanks for the description, @e0. Now I understand why the reactants were named "outgoing" metabolites.

I agree with @haowang-bioinfo - if possible, it would be nice to just use the stoichiometry values. Typically the stoichiometric coefficients are used to store information about which side of the reaction the metabolites are on, where positive = product and negative = reactant.

But if this cannot be done for some reason, then your suggestion to explain it clearly in the API documentation would also be an ok fix.

**Comment by e0** [comment]
Thank you for the feedback! It's always nice to understand things more from a researcher's perspective. I have moved this issue to the top of our backlog refinement list so that we can discuss it as a group soon.

I have a follow up question, and please let me if the suggestion is too overkill. **For bidirectional reactions, would it still be confusing if some metabolites have negative `stoichiometry` values?** It seems like it could be a similar problem if a negative value is supposed to represent a reactant. I feel like it would be the least confusing if the reaction could have their metabolites divided into two sides, for example something like this:

- reaction1
  - bidirectional: `true`
  - metabolites
    - left:
      - metabolite1
        - stoichiometry: `2`
    - right:
      - metabolite2
        - stoichiometry: `1`
      - metabolite3
        - stoichiometry: `1`

This way for unidirectional (`bidirectional`: `false`) reactions, `left` would always contain the reactants. For bidirectional (`bidirectional`: `true`) reactions, there would be no source of confusion about which metabolites "look like" reactants.

Anyhow, for backwards compatibility reasons, I think we need to keep the `outgoing` value for now and add clear documentation about it. For further API versions we should try to remove usages of `outgoing` altogether.

**Comment by haowang-bioinfo** [comment]
> For bidirectional reactions, would it still be confusing if some metabolites have negative stoichiometry values?

The negative stoichiometry values in a GEM mean that these metabolites are to be placed in the left side of the equations, regardless of bidirectional or unidirectional status. Therefore, it would be nice to explicitly provide reversibility status (bidirectional or unidirectional) in API responses, as the replacement of "outgoing".

**Comment by mihai-sysbio** [comment]
> Anyhow, for backwards compatibility reasons, I think we need to keep the `outgoing` value for now and add clear documentation about it.

The way it looks now, `outgoing` is not an immediately usable field, and would therefore be in favour of removing it, as previously suggested. @e0 could you please detail more the backwards compatibility concerns?

A complementary perspective is that documentation can be also a suboptimal fix. Having to explain confusing naming in documentation, like in this case, is a band-aid and not a robust fix.

Also please make sure the updated API response for `related-reactions` is fully aligned with the `reaction` endpoint.

**Comment by e0** [comment]
> The way it looks now, `outgoing` is not an immediately usable field, and would therefore be in favour of removing it, as previously suggested. @e0 could you please detail more the backwards compatibility concerns?

I don't know the real world usages of API endpoints. The standard rules for API is that it should not break. For example, if someone is building something on top of the API and relies on the `outgoing` field, their solution could now break if we were to remove it without warning.

**Comment by haowang-bioinfo** [comment]
@e0 I agree that it's indeed hard to know the actual usage of API. While the response from real user shows that confusion has been built upon the `outgoing` field, whose removal could then "break" further misunderstanding.

**Comment by e0** [comment]
@mihai-sysbio This is not urgent but when you have time please let us know how we should move forward with this issue. I have given the recommendation above but we are happy with whatever decision that makes the most sense for our project.
